File: cargo/crates/numbers-utils/tests/operations/factorial_utils_tests.rs
use chrono::Local;
use colored::Colorize;

use num_bigint::BigUint;
use numbers_utils::operations::factorial_utils::{factorial, factorial_random};

fn format_error_message(expected: &BigUint, actual: &BigUint) -> colored::ColoredString {
    format!(
        "\n\nTest run at: {}\nExpected : {}\nActual   : {}\n",
        Local::now().to_rfc3339(),
        expected,
        actual
    )
    .red()
}

#[test]
fn factorial_test() {
    let n = 75;
    let expected_output = BigUint::parse_bytes(b"24809140811395398091946477116594033660926243886570122837795894512655842677572867409443815424000000000000000000", 10).unwrap();
    let actual_output = factorial(n);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

#[test]
fn factor_random_test() {
    let result = factorial_random();
    let start_string = "Fact[";
    let error_message = format!("\nExpected: {}\n     Got: {}\n", start_string, result).red();
    assert!(result.starts_with(start_string), "{}", error_message,);
    assert!(result.contains(" = "));
}

File: cabal/packages/numbers-utils/tests/Operations/FactorialUtilsTests.hs
{-# LANGUAGE OverloadedStrings #-}

module Operations.FactorialUtilsTests (factorialTest, factorialRandomTest) where

import Data.List (isInfixOf, isPrefixOf)

import Data.Time.Clock (getCurrentTime)
import Data.Time.LocalTime (utcToLocalZonedTime)
import Test.Tasty (TestTree)
import Test.Tasty.HUnit (assertBool, testCase, assertEqual)

import Numbers.Utils.Operations.FactorialUtils (factorial, factorialRandom)

formatErrorMessage :: String -> String -> IO String
formatErrorMessage expected actual = do
    now <- getCurrentTime >>= utcToLocalZonedTime
    return $
        "\nTest run at: "
            ++ show now
            ++ "\nExpected : "
            ++ expected
            ++ "\nActual   : "
            ++ actual
            ++ "\n"

factorialTest :: TestTree
factorialTest = testCase "Factorial" $ do
    let n = 75
    let expectedOutput =
            read
                "24809140811395398091946477116594033660926243886570122837795894512655842677572867409443815424000000000000000000"
                :: Integer
    let actualOutput = factorial n
    errorMessage <- formatErrorMessage (show expectedOutput) (show actualOutput)
    assertEqual errorMessage expectedOutput actualOutput

factorialRandomTest :: TestTree
factorialRandomTest = testCase "Factorial Random Test" $ do
    result <- factorialRandom
    let startString = "Fact["
    let errorMessage = "\nExpected: " ++ startString ++ "\n     Got: " ++ result ++ "\n"
    assertBool errorMessage (startString `isPrefixOf` result)
    assertBool errorMessage (" = " `isInfixOf` result)
